packages/material-ui/src/Button/Button.js

```diff
@@ -58,20 +58,30 @@ const ButtonRoot = styled(ButtonBase, {
   shouldForwardProp: (prop) => rootShouldForwardProp(prop) || prop === 'classes',
   name: 'MuiButton',
   slot: 'Root',
-  overridesResolver: (props, styles) => {
-    const { styleProps } = props;
+  overridesResolver: [
+    (props, styles) => styles.root,
+    ({ styleProps }, styles) => styles[styleProps.variant],
+    ({ styleProps }, styles) => styles[`${styleProps.variant}${capitalize(styleProps.color)}`],
+    ({ styleProps }, styles) => styles[`size${capitalize(styleProps.size)}`],
+    ({ styleProps }, styles) => styles[`${styleProps.variant}Size${capitalize(styleProps.size)}`],
+    ({ styleProps }, styles) => styleProps.color === 'inherit' && styles.colorInherit,
+    ({ styleProps }, styles) => styleProps.disableElevation && styles.disableElevation,
+    ({ styleProps }, styles) => styleProps.fullWidth && styles.fullWidth,
+  ],
```

Comment:
While we are at it, maybe we should return a path, instead of the styles? 
```suggestion
    'root',
    styleProps.variant,
    `${styleProps.variant}${capitalize(styleProps.color)}`,
    `size${capitalize(styleProps.size)}`,
    `${styleProps.variant}Size${capitalize(styleProps.size)}`,
    styleProps.color === 'inherit' && 'colorInherit',
    styleProps.disableElevation && 'disableElevation',
    styleProps.fullWidth && 'fullWidth',
  ],
```

Reply:
how do you get the `styleProps`, it should return array of function.

Reply:
We could get it from overridesResolver.props.styleProps. The value would be about 1. reducing boilerplate and 2. allowing to supports accepting props in the theme definions in the future if we decide too (I don't see why we wouldn't but maybe we won't).

Reply:
for reducing boilerplate, I think this can help (I don't think we need styles).
```js
[
  () => 'root',
  ({ styleProps }) => styleProps.variant,
  ({ styleProps }) => `${styleProps.variant}${capitalize(styleProps.color)}`,
  ({ styleProps }) => `size${capitalize(styleProps.size)}`,
  ({ styleProps }) => `${styleProps.variant}Size${capitalize(styleProps.size)}`,
  ({ styleProps }) => styleProps.color === 'inherit' && 'colorInherit',
  () => 'disableElevation',
  ({ styleProps }) => styleProps.fullWidth && 'fullWidth',
]
```
instead of return `styles.[...]`, just return the key and createStyled will take care of it.

from your comment, I still don't see how to get the props at one place and fix the bug.

Reply:
@mnajdova what do you think about the approach

Reply:
@siriwatknp the idea was around doing something like:

```jsx
({ styleProps }) => [
  'root',
   styleProps.variant,
  `${styleProps.variant}${capitalize(styleProps.color)}`,
   size${capitalize(styleProps.size)}`,
 ...
```

Reply:
> and then have the styled() API so its magic to resolve the values and merge them correctly.

👍 plus one on this. The `styled()` API already accepts array, maybe we should just leverage it

Reply:
> at the component level, and then have the styled() API so its magic to resolve the values and merge them correctly.

+1 one function should be enough, you could just adjust the logic here https://github.com/mui-org/material-ui/pull/26824/files#diff-b8ae117b397ff9b4a1aef7dc493b4d8dcd79a2aec5a5a39c82c88fee5aebfd6aR121 to result first the result and check if the result is array. Would be great if you can ensure that this would work in all scenarios in case we are missing something, but I think it should be enough...

Reply:
@oliviertassinari I thought array would not work, but I checked in styled-components & emotion. Both works! 🎉 will update code tomorrow.